### PR TITLE
🚨🚨 Generate: change order of ops in beam sample to avoid nans

### LINKED ISSUE #26332: Beam Search Fails for Llama 70b
### System Info

- `transformers` version: 4.33.2
- Platform: Linux-5.15.0-1041-aws-x86_64-with-glibc2.35
- Python version: 3.10.12
- Huggingface_hub version: 0.17.2
- Safetensors version: 0.3.3
- Accelerate version: 0.23.0
- Accelerate config: 	not found
- PyTorch version (GPU?): 2.0.1+cu117 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: Yes, 6 A100 GPUs
- Using distributed or parallel set-up in script?: It is device_map="auto" which I believe is distributing the layers between GPUs

### Who can help?

@gante Appears to the be the relevant developer because this is an issue with model.generate

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

The below script only needs a change to have a TOKEN and CACHE_DIR and can be ran to generate the error.
I have tried with and without autocast, and it does not affect this.
I have also verified that the GPUs/Machine are not memory constrained.
Greedy generation works as expected, only beam-search is failing.
```
import os
import sys
from copy import deepcopy
import pandas as pd
import torch
from tqdm import tqdm

sys.path.append("/app")
from src.secrets import TOKEN
from src.constants import CACHE_DIR
from transformers import AutoModelForCausalLM, AutoTokenizer

tqdm.pandas()
# Use 6/8 A100 GPUs 
os.environ["CUDA_VISIBLE_DEVICES"] = "2,3,4,5,6,7"

# Load the model
path = "meta-llama/Llama-2-70b-chat-hf"
model = AutoModelForCausalLM.from_pretrained(path, cache_dir=CACHE_DIR, device_map="auto", torch_dtype=torch.float16, use_auth_token=TOKEN, use_cache=True)
# Setup Tokenizer
tokenizer = AutoTokenizer.from_pretrained(path, truncation_side="left", token=TOKEN)

prompts = [
    "<s>[INST]Hi, how are you?[/INST]",
    "<s>[INST]Who is the president?[/INST]",
    "<s>[INST]What continent is Ireland in?[/INST]",
    "<s>[INST]How fast can a chicken run?[/INST]",
    
]

for prompt in prompts:
    input_ids = tokenizer(prompt, return_tensors="pt", truncation=True, max_length=4096, add_special_tokens=False).input_ids
    # Move tokens to GPU
    input_ids = input_ids.to("cuda")
    with torch.no_grad():
        with torch.cuda.amp.autocast():
        output = model.generate(input_ids, num_beams=4, max_new_tokens=512, temperature=0.5, top_p=0.9)
    gen_text = tokenizer.batch_decode(output)
    print(gen_text[0])
```

The resulting error
```
RuntimeError                              Traceback (most recent call last)
Cell In[5], line 7
      4 input_ids = input_ids.to("cuda")
      5 with torch.no_grad():
      6     #with torch.cuda.amp.autocast():
----> 7     output = model.generate(input_ids, num_beams=4, max_new_tokens=512, temperature=0.5, top_p=0.9)
      8 gen_text = tokenizer.batch_decode(output)
      9 print(gen_text[0])

File /usr/local/lib/python3.10/dist-packages/torch/utils/_contextlib.py:115, in context_decorator.<locals>.decorate_context(*args, **kwargs)
    112 @functools.wraps(func)
    113 def decorate_context(*args, **kwargs):
    114     with ctx_factory():
--> 115         return func(*args, **kwargs)

File /usr/local/lib/python3.10/dist-packages/transformers/generation/utils.py:1718, in GenerationMixin.generate(self, inputs, generation_config, logits_processor, stopping_criteria, prefix_allowed_tokens_fn, synced_gpus, assistant_model, streamer, negative_prompt_ids, negative_prompt_attention_mask, **kwargs)
   1710     input_ids, model_kwargs = self._expand_inputs_for_generation(
   1711         input_ids=input_ids,
   1712         expand_size=generation_config.num_beams,
   1713         is_encoder_decoder=self.config.is_encoder_decoder,
   1714         **model_kwargs,
   1715     )
   1717     # 14. run beam sample
-> 1718     return self.beam_sample(
   1719         input_ids,
   1720         beam_scorer,
   1721         logits_processor=logits_processor,
   1722         logits_warper=logits_warper,
   1723         stopping_criteria=stopping_criteria,
   1724         pad_token_id=generation_config.pad_token_id,
   1725         eos_token_id=generation_config.eos_token_id,
   1726         output_scores=generation_config.output_scores,
   1727         return_dict_in_generate=generation_config.return_dict_in_generate,
   1728         synced_gpus=synced_gpus,
   1729         **model_kwargs,
   1730     )
   1732 elif generation_mode == GenerationMode.GROUP_BEAM_SEARCH:
   1733     # 11. prepare beam search scorer
   1734     beam_scorer = BeamSearchScorer(
   1735         batch_size=batch_size,
   1736         num_beams=generation_config.num_beams,
   (...)
   1742         max_length=generation_config.max_length,
   1743     )

File /usr/local/lib/python3.10/dist-packages/transformers/generation/utils.py:3392, in GenerationMixin.beam_sample(self, input_ids, beam_scorer, logits_processor, stopping_criteria, logits_warper, max_length, pad_token_id, eos_token_id, output_attentions, output_hidden_states, output_scores, return_dict_in_generate, synced_gpus, **model_kwargs)
   3388 next_token_scores = next_token_scores.view(batch_size, num_beams * vocab_size)
   3390 probs = nn.functional.softmax(next_token_scores, dim=-1)
-> 3392 next_tokens = torch.multinomial(probs, num_samples=2 * num_beams)
   3393 next_token_scores = torch.gather(next_token_scores, -1, next_tokens)
   3395 next_token_scores, _indices = torch.sort(next_token_scores, descending=True, dim=1)

RuntimeError: probability tensor contains either `inf`, `nan` or element < 0
```

### Expected behavior

The model generates tokens using beam search.

### PR DESCRIPTION
**Pull request by gante**
# What does this PR do?

Fixes https://github.com/huggingface/transformers/issues/26332 -- see [this](https://github.com/huggingface/transformers/issues/26332#issuecomment-1764736970) comment for a full explanation.

TL;DR: in `beam_sample`, `logits_warper` are now applied BEFORE adding the beam scores. We have been postponing this change to avoid introducing output differences, but the truth is that the order of operations causes issues (e.g. [1](https://github.com/huggingface/transformers/issues/26332) [2](https://github.com/huggingface/transformers/issues/22914))

This is technically a bug fix (`beam_sample` is unusable for long generations with `temperature < 1.0` before this change), hence the lack of a deprecation cycle. However, it may alter some generated outputs (short generations with low temperature), hence the 🚨🚨.

Please note that other common operators, like `top_p` and `top_k`, are unaltered by this change.

### CODE REVIEW

File: src/transformers/generation/utils.py [hunk 1]
```diff
@@ -820,11 +820,20 @@ def _get_logits_warper(
         # instantiate warpers list
         warpers = LogitsProcessorList()
 
+        # In beam methods, we need to keep at least one non-eos token to explore continuations that might have a
+        # better score (i.e. keep len(generation_config.eos_token_id) + 1)
```
Comment (ArthurZucker) [inline src/transformers/generation/utils.py:824]: it's not evidend that generation_config is a list as it usually is an int no? 
```suggestion
        # In beam methods, we need to keep at least one non-eos token to explore continuations that might have a
        # better score (i.e. keep len(list(generation_config.eos_token_id)) + 1)
```

File: src/transformers/generation/utils.py [hunk 2]
```diff
@@ -820,11 +820,20 @@ def _get_logits_warper(
...
+        if generation_config.num_beams > 1:
+            if isinstance(generation_config.eos_token_id, list):
+                min_tokens_to_keep = len(generation_config.eos_token_id) + 1
+            else:
+                min_tokens_to_keep = 2
```
Comment (ArthurZucker) [inline src/transformers/generation/utils.py:829]:
```suggestion
            min_tokens_to_keep = len(list(generation_config.eos_token_id)) + 1
```
should work as
```python 
>>> list([1,2])
[1,2]
```

File: src/transformers/generation/tf_utils.py [hunk 3]
```diff
@@ -1430,14 +1430,22 @@ def _get_logits_warper(
...
 
-        # the following idea is largely copied from this PR: https://github.com/huggingface/transformers/pull/5420/files
-        # all samplers can be found in `generation_utils_samplers.py`
+        # In beam methods, we need to keep at least one non-eos token to explore continuations that might have a
+        # better score (i.e. keep len(generation_config.eos_token_id) + 1)
+        if generation_config.num_beams > 1:
+            if isinstance(generation_config.eos_token_id, list):
+                min_tokens_to_keep = len(generation_config.eos_token_id) + 1
+            else:
+                min_tokens_to_keep = 2
```
Comment (ArthurZucker) [inline src/transformers/generation/tf_utils.py:1439]:
```suggestion
            min_tokens_to_keep = len(list(generation_config.eos_token_id)) + 1
```

**Review by ArthurZucker — APPROVED** [review]
Left small comments, LGTM

File: src/transformers/generation/tf_utils.py (hunk 3 above)
Comment (gante) [reply src/transformers/generation/tf_utils.py:1439]: The catch is that `generation_config.eos_token_id` can be `None`, and `len(list(None))` throws `TypeError: 'NoneType' object is not iterable` :(

Comment (ArthurZucker) [reply src/transformers/generation/tf_utils.py:1439]: Ah right 💯
